Extensions/ArtifactEngine/Engine/artifactEngine.ts

```diff
@@ -149,9 +190,84 @@ export class ArtifactEngine {
+            if(!artifactEngineOptions.enableIncrementalDownload) {
+                this.newHashMap = {};
+                resolve();
+            }
+            else {
+                sourceProvider.getArtifactItems(itemsToProcess[0]).then((items: models.ArtifactItem[]) => {
+                    sourceProvider.getArtifactItem(items.find(x => path.normalize(x.path) === path.join(itemsToProcess[0].path, 'artifact-metadata.csv'))).then((hashStream : NodeJS.ReadableStream) => {                        
+                        var newHashPromise = new Promise((resolve) => {
+                            var newHash = readline.createInterface ({
+                                input : hashStream
+                            });
+                    
+                            newHash.on('line', (line) => {
+                                var words = line.split(',');
+                                this.newHashMap[words[0]] = words[1];
+                            });
+                
+                            newHash.on('close', () => {
+                                resolve();
+                            });
+                        });
+                        newHashPromise.then(() => {
+                            this.artifactItemStore.setHashMap(this.newHashMap)
+                            resolve();                                   
+                        });                            
+                    }, (err) => {
+                        Logger.logInfo("Incremental Download Failed. Downloading normally.")
+                        artifactEngineOptions.enableIncrementalDownload = false;
+                        this.newHashMap = {};
+                        resolve();
+                    });
+                }, (err) => {
+                    Logger.logInfo("Incremental Download Failed. Downloading normally.")
+                    artifactEngineOptions.enableIncrementalDownload = false;
+                    this.newHashMap = {};
+                    resolve();
+                });
+            }
+        });
+    }
+
+    cacheUpdation(sourceProvider: models.IArtifactProvider, destProvider: models.IArtifactProvider, resolve) {    
+        if(this.artifactItemStore._downloadTickets.find(x => x.state === models.TicketState.Failed)) {
+            sourceProvider.dispose();
+            destProvider.dispose();
+            resolve(this.artifactItemStore.getTickets());
+        }
+        else {
+            var destination = destProvider.getRootLocation();
+            var artifactName = sourceProvider.getRootItemPath();                        
+            var cachePath = this.cacheProvider.getCacheDirectory();
+            if(fs.existsSync(cachePath)) {
+                tl.rmRF(cachePath);
+            }
+            var self = this;
+            var cacheValidator = false;
+            tl.mkdirP(cachePath);
+            tl.cp((path.join(destination,artifactName)+'/.'), cachePath, '-r');
```

Comment:
path.join instead of +

Reply:
cannot use path.join() as it removes the trailing dot(.) which is necessary in this case because without it, another directory is made inside the destination location